File: NET/LatencyTesting/LatencyTesting_CSharp/EffectBase.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace LatencyTesting
{
    public abstract class EffectBase
    {
        private EffectBase decorate;

        public void Process(float[] buffer)
        {
            Transform(buffer);
            if (decorate != null)
            {
                decorate.Process(buffer);
            }
        }

        public EffectBase Decorator
        {
            set 
            {
                if (decorate == null)
                {
                    decorate = value;
                }
                else
                {
                    decorate.Decorator = value;
                }
            }
        }

        public abstract void Setup(EffectConfiguration config);
        protected abstract void Transform(float[] buffer);
    }
}

File: NET/LatencyTesting/LatencyTesting_C++/EffectBase.cpp
#include "stdafx.h"
#include "EffectBase.h"

namespace Processor
{
	EffectBase::EffectBase()
	{
		_decorate = nullptr;
	}
	
	void EffectBase::Process(array<float>^ buffer)
	{
		Transform(buffer);
		if(_decorate != nullptr)
		{
			_decorate->Process(buffer);
		}
	}

	void EffectBase::Decorator::set(EffectBase^ value)
	{
		if(_decorate == nullptr)
		{
			_decorate = value;
		}
		else
		{
			_decorate->Decorator = value;
		}
	}
}